### PR TITLE
Add taiko drumroll skinning support

### PR DESCRIPTION
**Pull request by peppy**
Still needs metric elements added, but mostly works.

- Adds remaining skinnable hitobjects (without implementations) to allow for using `SkinnableDrawable` across `TaikoHitObject` classes.
- Tidies up sizing logic (aka removes a heap that we don't need)
- Stretched piece gets gradiented to transparent at each end. Will fix this in a follow-up PR (also happens in mania, see https://github.com/ppy/osu/issues/8543)
- [x] Figure why LegacyCirclePiece is not visible after rewind (related to proxying?)

### CODE REVIEW

File: osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableDrumRollTick.cs [hunk 1]
```diff
@@ -20,10 +20,8 @@ public DrawableDrumRollTick(DrumRollTick tick)
 
         public override bool DisplayResult => false;
 
-        protected override CompositeDrawable CreateMainPiece() => new TickPiece
-        {
-            Filled = HitObject.FirstTick
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableDrumRollTick.cs:25]: Removed?

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Comment by bdach** [comment]
On a quick-and-dirty check the circle piece is indeed visible after rewind if it's proxied upon hit.

On another note, from what I can see the current implementation has `LegacyCirclePiece` take the width of 1, which brings back #6889 in the case of legacy skins, as the "fix" I applied once (calculating offset of drawable origin from its visible edge - which broke once already in catch) also stops working on the drum roll because of that ([video](https://streamable.com/2p8w9b)).

**Comment by peppy** [comment]
Yeah, I figured there was probably a reason for that being the way it was. Guess I'll have to somehow hack it into the draw space.

**Comment by peppy** [comment]
@bdach in your video the drumrolls are also offset incorrectly temporally. i can't reproduce this, were you on the latest commit of this branch?

**Comment by bdach** [comment]
I'm somewhat sure I did - my head of this branch is at bfc0d41, but I messed with the code a little bit so I can't recall now if the tree was clean or not. Will try to repro and provide details later today.

File: osu.Game.Rulesets.Taiko/Objects/Drawables/Pieces/ElongatedCirclePiece.cs [hunk 2]
```diff
@@ -12,18 +14,15 @@ public ElongatedCirclePiece()
...
         {
             base.Update();
-
-            var padding = Content.DrawHeight * Content.Width / 2;
-
-            Content.Padding = new MarginPadding
-            {
-                Left = padding,
-                Right = padding,
-            };
```
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Objects/Drawables/Pieces/ElongatedCirclePiece.cs:25]: Can confirm reproduction of the misalignment; it shows up even on default lazer skin in this PR (or, to be more precise, on f36477e3 with a clean tree), doesn't show up on `master`. Will be visible in `TestSceneAutoplay`. Screenshots:

| `master` | this PR |
| :-: | :-: |
| image | image |

Seems to be related to removing the padding as highlighted. Bringing it back fixes the alignment of `ElongatedCirclePiece` itself but ticks remain broken (too stretched out).

Comment (peppy) [reply osu.Game.Rulesets.Taiko/Objects/Drawables/Pieces/ElongatedCirclePiece.cs:25]: The padding needs a local comment explaining exactly what it's doing until we fix this in a better way. I thought I was able to remove it due to the refactors I made and there was nothing noticeably failing as a result.

Comment (peppy) [reply osu.Game.Rulesets.Taiko/Objects/Drawables/Pieces/ElongatedCirclePiece.cs:25]: @bdach can you explain how the offset is occurring? in your screenshots you show it changing at the `DrawableDrumRoll` level, but I can't seem to find any changes I've made which should affect that.

Comment (bdach) [reply osu.Game.Rulesets.Taiko/Objects/Drawables/Pieces/ElongatedCirclePiece.cs:25]: Not currently in a position to investigate in detail but from what I remember the actual drumroll's position hasn't changed from `master`, it's the `ElongatedCirclePiece` lower down that starts protruding out of it, breaking alignment.

Comment (peppy) [reply osu.Game.Rulesets.Taiko/Objects/Drawables/Pieces/ElongatedCirclePiece.cs:25]: I think I've fixed this by applying at a more central level. Can you check on the solution?

Comment (bdach) [reply osu.Game.Rulesets.Taiko/Objects/Drawables/Pieces/ElongatedCirclePiece.cs:25]: Give me a few hours and I'll get back to you on that.

Comment (bdach) [reply osu.Game.Rulesets.Taiko/Objects/Drawables/Pieces/ElongatedCirclePiece.cs:25]: Took me a little longer than I promised, but as far as I can see the alignment is fixed, although this set:

https://github.com/ppy/osu/blob/58a1c6e17186084345feedbd1c5b0d44c2ce8695/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableDrumRoll.cs#L145

doesn't seem to have any effect as the getter is hard-overridden in `DrawableTaikoHitObject` itself:

https://github.com/ppy/osu/blob/8b83c3daabf83075ce0be0c47948a312529c9e1d/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableTaikoHitObject.cs#L113

so the value gets set but never actually read (draw visualiser shows the one from base).

**Comment by bdach** [comment]
If you haven't figured it out yet, I looked into this some more and as for the reason that proxying helps for visibility of the circle piece at the start, I found out that once it scrolls off-screen it becomes masked away, which xmldocs indicate proxying would help with:

https://github.com/ppy/osu/blob/8b83c3daabf83075ce0be0c47948a312529c9e1d/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableTaikoHitObject.cs#L38-L41

One piece of the puzzle is that `DrawableHitObject` itself prevents updates of masking before all nested objects are judged:

https://github.com/ppy/osu/blob/9dfca4a374a20ac0aea6ca8d62fd6d41818e72b3/osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs#L401

which prevents the masking from being reprocessed after rewind (removing the `AllJudged` check also makes the circle piece appear).

As to why only the circle piece at the start gets this treatment, I'm almost sure this is due to the positioning of the elements themselves. The circle piece is the only element of the drum roll that has a chance to be masked away before the entire drum roll is, so it gets set `false` when it goes off-screen. When the *entire* drum roll goes off-screen, however, it gets masked away *itself*, therefore not propagating masking downwards (`UpdateSubTreeMasking` early-returns and never sets `IsMaskedAway = true` on children).

**Comment by smoogipoo** [comment]
I only allowed masking to continue when `AllJudged = true` just to enable optimisations. It wasn't fixing an issue or anything.

Given that we've tightened up lifetime quite a lot, I'd be fine with removal of the `AllJudged` component (`UpdateSubTreeMasking() => false`).

**Comment by peppy** [comment]
Looks good with that change.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**